llap-client/src/java/org/apache/hadoop/hive/llap/tezplugins/helpers/LlapTaskUmbilicalServer.java

```diff
@@ -54,27 +56,54 @@
 
   public LlapTaskUmbilicalServer(Configuration conf, LlapTaskUmbilicalProtocol umbilical, int numHandlers) throws IOException {
     jobTokenSecretManager = new JobTokenSecretManager();
-    int umbilicalPort = HiveConf.getIntVar(conf, HiveConf.ConfVars.LLAP_TASK_UMBILICAL_SERVER_PORT);
-    if (umbilicalPort <= 0) {
-      umbilicalPort = 0;
+
+    String[] portRange =
+        conf.get(HiveConf.ConfVars.LLAP_TASK_UMBILICAL_SERVER_PORT.varname)
+            .split("-");
+
+    int minPort = Integer.parseInt(portRange[0]);
+    boolean portFound = false;
+    IOException e = null;
+    if (portRange.length == 1) {
+      // Single port specified, not Range.
+      startServer(conf, umbilical, numHandlers, minPort);
+      portFound = true;
+    } else {
+      int maxPort = Integer.parseInt(portRange[1]);
+      for (int i = minPort; i < maxPort; i++) {
+        try {
+          startServer(conf, umbilical, numHandlers, i);
+          portFound = true;
+          break;
+        } catch (BindException be) {
+          // Ignore and move ahead, in search of a free port.
```

Comment:
Log at warn level to say which port is being tried and what error message received for debugging.